File: CoinChange.swift
// Time Complexity : (O)2n^m . n = number of coins. m = amount
// Space Complexity : (O)2n^m
// Did this code successfully run on Leetcode : timelimit exceeded


//Recursive solution
//Make two decision and get a number of coins needed to make the given amount
//choose a coin or not choose a coins
//then compare the min of two result and return the minimum

class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        //edge
        if coins.count == 0 {
            return 0
        }
        return helper(coins,amount,0,0)
    }
    
    func helper(_ coins: [Int],_ amount: Int, _ index: Int,_ min: Int) -> Int {
        //base
        if amount == 0 {
            return min
        }
        
        if amount < 0 || index == coins.count {
            return -1
        }
        //logic
        //choose a coin
        let case1: Int = helper(coins, amount - coins[index], index, min+1)
        // don't choose a coin
        let case2: Int = helper(coins, amount, index+1, min)
        
        if case1 == -1 {
            return case2
        }
        if case2 == -1 {
            return case1
        }
        
        Swift.min(case1,case2)
    }
}


//DP tabular solution
//Create a matrix to find all minimum number of coins to make a given amount from zero
//Create 2d array with first row infinity and first column zero
//set minimum number of coins needed to make the amount. There are two cases. case 1 is not to select a coin and case 2 is select a coin. For not select a coin get the previous(row) minimum value. For select a coin find the previously calculated minimum of coin to make a amount less than the current row amount.
//Don't need to slect a coin if coin is greater than row amount
//get min of zero case and selected case value

class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        //edge
        if coins.count == 0 {
            return 0
        }
        var dp = [[Int]](repeating: [Int](repeating: 0, count: amount+1), count: coins.count+1)

        //fill infinity for first row
        for index in 1..<dp[0].count {
            dp[0][index] = 9999
        }
        //fill zero for first dummy column
        for index in 0..<dp.count  {
            dp[index][0] = 0
        }
        for i in 1..<dp.count { // amount
            for j in 1..<dp[0].count { // coin
                if j < coins[i - 1] {
                    dp[i][j] = dp[i-1][j]
                } else {
                    dp[i][j] = Swift.min(dp[i-1][j], dp[i][j - coins[i-1]] + 1) //min of zero case and selected case value
                }
            }
        }
        let result = dp[coins.count][amount]
        return result >= 9999 ? -1 : result
    }
    
}

File: src/com/madhurima/CoinChange.java
/*
    Time Complexity = O(amount * coins)
    Space Complexity = O(amount * coins)
    Did this code successfully run on Leetcode : yes
*/

package com.madhurima;

public class CoinChange {
}

class SolutionA1 {
    public int coinChange(int[] coins, int amount) {

        if(coins == null || coins.length == 0 || amount < 0){
            return -1;
        }

        int n = coins.length;
        int[][] dp = new int[n+1][amount + 1];

        //fill first row from 2nd column i.e. from i=1 with infinity i.e. amount+1
        for(int i = 1; i < amount+1; i++){
            dp[0][i] = amount+1;
        }

        for(int i = 1; i < n+1; i++){
            for(int j = 1; j < amount + 1; j++){
                if(j < coins[i-1]){
                    dp[i][j] = dp[i-1][j];
                }else{
                    dp[i][j] = Math.min(dp[i-1][j], dp[i][j-coins[i-1]] + 1);
                }

            }
        }

        if(dp[n][amount] == amount+1){
            return -1;
        }

        return dp[n][amount];

        // return helper(coins, amount, 0, 0);

    }

    //Recursion helper function
    private int helper(int[] coins, int amount, int idx, int coinsUsed){

        if(idx == coins.length || amount < 0){
            return -1;
        }

        if(amount == 0){
            return coinsUsed;
        }

        //case zero: not chosen
        int case1 = helper(coins, amount, idx+1, coinsUsed);

        //case one: chosen
        int case2 = helper(coins, amount - coins[idx], idx, coinsUsed+1);

        if(case1 == -1){
            return case2;
        }

        if(case2 == -1){
            return case1;
        }

        return Math.min(case1, case2);
    }

    public static void main(String[] args) {
        SolutionA1 s = new SolutionA1();
        int[] coins = {1,2,5};
        int target = 11;
        System.out.println(s.coinChange(coins, target));
    }


}
